### PR TITLE
build.d: Add --dry-run option

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
Inspired by GNU Make's `--dry-run` flag to ease development and maintenance of `build.d`

CC @marler8997

### CODE REVIEW

File: src/build.d [hunk 1]
```diff
@@ -281,10 +283,11 @@ alias versionFile = memoize!(function()
     Dependency dep;
     with (dep)
     {
+        name = "versionFile";
```
Comment (marler8997) [inline src/build.d]: Not every target needs a name.  Without the name, the help just shows the file that this target generates which in this case I think i s better.

File: src/build.d [hunk 2]
```diff
@@ -313,11 +316,12 @@ alias sysconfDirFile = memoize!(function()
     Dependency dep;
     with(dep)
     {
+        name = "sysconfDirFile";
```
Comment (marler8997) [inline src/build.d]: Same comment here.

File: src/build.d [hunk 3]
```diff
@@ -1178,12 +1182,23 @@ class DependencyRef
         if (msg)
             msg.writeln;
 
-        if (commandFunction !is null)
-            return commandFunction();
-
-        if (command)
+        if(dryRun)
+        {
+            if(commandFunction)
```
Comment (marler8997) [inline src/build.d]: You could try a set of criteria on what to print.  For example, if `name` is null, you could check if there are any outputs and print `generating %s` instead.

Comment (marler8997) [reply src/build.d]: We could also just make it an error if `commandFunction` is set without `msg` field being set.

Comment (MoonlightSentinel) [reply src/build.d]: Changed to list the dependencies `name` and `targets` and fall back to `Executing commandFunction()`.

I think every dependency **must** have a `msg` to reliable track why a build failed.

**Review by marler8997 — APPROVED** [review]
Output looks nice

**Comment by marler8997** [comment]
I had an afterthought.  Note that all of this is run in parallel.  So the output could become a bit screwed up, might be better to build a single string for each target and call write one time.

**Comment by MoonlightSentinel** [comment]
Interesting thought, but i tend to avoid parallel builds while working on build scripts. This results in a clear visualisation of the dependency graph.

But i can change it just to be sure.

**Review by wilzbach — APPROVED** [review]
Nice idea!

**Pull request merged**